procmon_parser/symbol_resolver/symbol_resolver.py

```diff
@@ -0,0 +1,711 @@
+
+            with p.open("rb") as f:
+                log_reader = ProcmonLogsReader(f, should_get_stacktrace=True)
+                sym_resolver = SymbolResolver(log_reader)
+                for i, event in enumerate(log_reader):
+                    print(f"{i:04x} {event!r}")
+                    frames = list(sym_resolver.resolve_stack_trace(event))
+                    print(StackTraceInformation.prettify(frames))
+            ```
+
+        Yields:
+            An instance of `StackTraceFrameInformation` for each of the frame in the stack trace.
+        """
+        if not event.stacktrace or event.stacktrace is None:
+            raise RuntimeError("Trying to resolve a stack trace while there is no stack trace.")
+
+        # Initialize dbghelp symbolic information.
+        pid = event.process.pid
+        self._dbghelp.SymInitialize(pid, None, False)
+
+        # set up callback if we are in debug mode
+        if self._debug_callback:
+            callback = PSYMBOL_REGISTERED_CALLBACK64(self._symbol_registered_callback)
+            self._dbghelp.SymRegisterCallbackW64(pid, callback, PVOID(pid))
+
+        logger.debug("# Stack Trace frames: {len_event_stack_trace}".format(
+            len_event_stack_trace=len(event.stacktrace)))
+        logger.debug("PID: {pid:#08x}".format(pid=pid))
+
+        # Resolve each of the addresses in the stack trace, frame by frame.
+        for frame_number, address in enumerate(event.stacktrace):
+            frame_type = FrameType.from_address(address, self._max_user_address)
+            logger.debug("{sep}\nStack Frame: {frame_number:04} type: {frame_type}".format(
+                sep='-' * 79, frame_number=frame_number, frame_type=frame_type))
+
+            # find the module that contains the given address. It might not be found.
+            logger.debug("Address: {address:#016x}".format(address=address))
+            module = self.find_module(event, address)
+            if not module:
+                yield StackTraceFrameInformation(frame_type, frame_number, address)
+                continue
+
+            logger.debug("Address: {address:#016x}  --> Module: {module!r}".format(address=address, module=module))
+
+            # We have the address and the module name. Get the corresponding file from the Symbol store!
+            # Once we have the file, we'll be able to query the symbol for the address.
+            found_file = ctypes.create_unicode_buffer(260 * 2)
+            module_id = PVOID(module.timestamp)
+            search_path = None  # use the default search path provided to SymInitialize.
+
+            # We give it two tries:
+            # 1. The module is an MS module, in which case it's going to be resolved pretty much automatically.
+            #   1.a if it's not an MS module it's going to fail.
+            # 2. If it's not an MS module, then we indicate to SymFindFileInPath where to find the binary in SearchPath.
+            for j in range(2):
```

Comment:
instead of the loop I think it would simplify to do something like

```python
def find_file():
    return self._dbghelp.SymFindFileInPathW(
        HANDLE(pid),  # hProcess
        search_path,  # SearchPath
        module.path,  # FileName (PCWSTR: it's fine to pass a python string)
        ctypes.byref(module_id),  # id
        module.size,  # two
        0,  # three
        SSRVOPT.SSRVOPT_GUIDPTR,  # flags: ProcMon uses 'SSRVOPT_GUIDPTR' but it's not a GUID??? still works
        found_file,  # [out] FoundFile
        PFINDFILEINPATHCALLBACK(0),  # callback (nullptr)
        None  # context
    )

ret_val = find_file()
if not ret_val:
    if ctypes.get_last_error() == 2:
        dir_path = pathlib.Path(module.path).parent
        search_path = str(dir_path) if dir_path.is_dir() else search_path
        ret_val = find_file()
```